File: test/sample-projects/go/user-management/internal/services/user_service.go
package services

import (
	"encoding/json"
	"errors"
	"fmt"
	"strings"
	"time"

	"github.com/example/user-management/internal/models"
	"github.com/example/user-management/internal/utils"
	"github.com/google/uuid"
	"gorm.io/gorm"
)

// UserService handles user-related business logic
type UserService struct {
	db *gorm.DB
}

// NewUserService creates a new user service
func NewUserService(db *gorm.DB) *UserService {
	return &UserService{db: db}
}

// CreateUser creates a new user
func (s *UserService) CreateUser(req *models.UserRequest) (*models.User, error) {
	// Check if username already exists
	var existingUser models.User
	if err := s.db.Where("username = ?", req.Username).First(&existingUser).Error; err == nil {
		return nil, errors.New("username already exists")
	}
	
	// Check if email already exists (if provided)
	if req.Email != "" {
		if err := s.db.Where("email = ?", req.Email).First(&existingUser).Error; err == nil {
			return nil, errors.New("email already exists")
		}
	}
	
	// Create new user
	user := &models.User{
		Role:   models.RoleUser,
		Status: models.StatusActive,
	}
	
	if err := user.FromRequest(req); err != nil {
		return nil, fmt.Errorf("failed to create user from request: %w", err)
	}
	
	if err := user.Validate(); err != nil {
		return nil, fmt.Errorf("user validation failed: %w", err)
	}
	
	if err := s.db.Create(user).Error; err != nil {
		return nil, fmt.Errorf("failed to create user: %w", err)
	}
	
	return user, nil
}

// GetUserByID retrieves a user by ID
func (s *UserService) GetUserByID(id uuid.UUID) (*models.User, error) {
	var user models.User
	if err := s.db.First(&user, "id = ?", id).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, errors.New("user not found")
		}
		return nil, fmt.Errorf("failed to get user: %w", err)
	}
	return &user, nil
}

// GetUserByUsername retrieves a user by username
func (s *UserService) GetUserByUsername(username string) (*models.User, error) {
	var user models.User
	if err := s.db.Where("username = ?", username).First(&user).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, errors.New("user not found")
		}
		return nil, fmt.Errorf("failed to get user: %w", err)
	}
	return &user, nil
}

// GetUserByEmail retrieves a user by email
func (s *UserService) GetUserByEmail(email string) (*models.User, error) {
	var user models.User
	if err := s.db.Where("email = ?", email).First(&user).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, errors.New("user not found")
		}
		return nil, fmt.Errorf("failed to get user: %w", err)
	}
	return &user, nil
}

// UpdateUser updates an existing user
func (s *UserService) UpdateUser(id uuid.UUID, updates map[string]interface{}) (*models.User, error) {
	user, err := s.GetUserByID(id)
	if err != nil {
		return nil, err
	}
	
	// Apply updates
	for key, value := range updates {
		switch key {
		case "name":
			if name, ok := value.(string); ok {
				user.Name = name
			}
		case "age":
			if age, ok := value.(int); ok {
				user.Age = age
			}
		case "email":
			if email, ok := value.(string); ok {
				user.Email = email
			}
		case "role":
			if role, ok := value.(models.UserRole); ok {
				user.Role = role
			}
		case "status":
			if status, ok := value.(models.UserStatus); ok {
				user.Status = status
			}
		case "metadata":
			if metadata, ok := value.(map[string]interface{}); ok {
				user.Metadata = metadata
			}
		}
	}
	
	if err := user.Validate(); err != nil {
		return nil, fmt.Errorf("user validation failed: %w", err)
	}
	
	if err := s.db.Save(user).Error; err != nil {
		return nil, fmt.Errorf("failed to update user: %w", err)
	}
	
	return user, nil
}

// DeleteUser soft deletes a user
func (s *UserService) DeleteUser(id uuid.UUID) error {
	user, err := s.GetUserByID(id)
	if err != nil {
		return err
	}
	
	user.Delete()
	
	if err := s.db.Save(user).Error; err != nil {
		return fmt.Errorf("failed to delete user: %w", err)
	}
	
	return nil
}

// HardDeleteUser permanently deletes a user
func (s *UserService) HardDeleteUser(id uuid.UUID) error {
	if err := s.db.Unscoped().Delete(&models.User{}, id).Error; err != nil {
		return fmt.Errorf("failed to hard delete user: %w", err)
	}
	return nil
}

// GetAllUsers retrieves all users with pagination
func (s *UserService) GetAllUsers(page, pageSize int) ([]*models.User, int64, error) {
	var users []*models.User
	var total int64
	
	// Count total users
	if err := s.db.Model(&models.User{}).Count(&total).Error; err != nil {
		return nil, 0, fmt.Errorf("failed to count users: %w", err)
	}
	
	// Get users with pagination
	offset := (page - 1) * pageSize
	if err := s.db.Limit(pageSize).Offset(offset).Find(&users).Error; err != nil {
		return nil, 0, fmt.Errorf("failed to get users: %w", err)
	}
	
	return users, total, nil
}

// GetActiveUsers retrieves all active users
func (s *UserService) GetActiveUsers() ([]*models.User, error) {
	var users []*models.User
	if err := s.db.Where("status = ?", models.StatusActive).Find(&users).Error; err != nil {
		return nil, fmt.Errorf("failed to get active users: %w", err)
	}
	return users, nil
}

// GetUsersByRole retrieves users by role
func (s *UserService) GetUsersByRole(role models.UserRole) ([]*models.User, error) {
	var users []*models.User
	if err := s.db.Where("role = ?", role).Find(&users).Error; err != nil {
		return nil, fmt.Errorf("failed to get users by role: %w", err)
	}
	return users, nil
}

// SearchUsers searches for users by name or username
func (s *UserService) SearchUsers(query string, page, pageSize int) ([]*models.User, int64, error) {
	var users []*models.User
	var total int64
	
	searchQuery := "%" + strings.ToLower(query) + "%"
	
	// Count total matching users
	if err := s.db.Model(&models.User{}).Where(
		"LOWER(name) LIKE ? OR LOWER(username) LIKE ? OR LOWER(email) LIKE ?",
		searchQuery, searchQuery, searchQuery,
	).Count(&total).Error; err != nil {
		return nil, 0, fmt.Errorf("failed to count search results: %w", err)
	}
	
	// Get matching users with pagination
	offset := (page - 1) * pageSize
	if err := s.db.Where(
		"LOWER(name) LIKE ? OR LOWER(username) LIKE ? OR LOWER(email) LIKE ?",
		searchQuery, searchQuery, searchQuery,
	).Limit(pageSize).Offset(offset).Find(&users).Error; err != nil {
		return nil, 0, fmt.Errorf("failed to search users: %w", err)
	}
	
	return users, total, nil
}

// GetUserStats returns user statistics
func (s *UserService) GetUserStats() (*utils.UserStats, error) {
	var stats utils.UserStats
	
	// Total users
	if err := s.db.Model(&models.User{}).Count(&stats.Total).Error; err != nil {
		return nil, fmt.Errorf("failed to count total users: %w", err)
	}
	
	// Active users
	if err := s.db.Model(&models.User{}).Where("status = ?", models.StatusActive).Count(&stats.Active).Error; err != nil {
		return nil, fmt.Errorf("failed to count active users: %w", err)
	}
	
	// Admin users
	if err := s.db.Model(&models.User{}).Where("role = ?", models.RoleAdmin).Count(&stats.Admin).Error; err != nil {
		return nil, fmt.Errorf("failed to count admin users: %w", err)
	}
	
	// Regular users
	if err := s.db.Model(&models.User{}).Where("role = ?", models.RoleUser).Count(&stats.User).Error; err != nil {
		return nil, fmt.Errorf("failed to count regular users: %w", err)
	}
	
	// Guest users
	if err := s.db.Model(&models.User{}).Where("role = ?", models.RoleGuest).Count(&stats.Guest).Error; err != nil {
		return nil, fmt.Errorf("failed to count guest users: %w", err)
	}
	
	// Users with email
	if err := s.db.Model(&models.User{}).Where("email != ''").Count(&stats.WithEmail).Error; err != nil {
		return nil, fmt.Errorf("failed to count users with email: %w", err)
	}
	
	return &stats, nil
}

// AuthenticateUser authenticates a user with username and password
func (s *UserService) AuthenticateUser(username, password string) (*models.User, error) {
	user, err := s.GetUserByUsername(username)
	if err != nil {
		return nil, errors.New("invalid username or password")
	}
	
	if !user.IsActive() {
		return nil, errors.New("user account is not active")
	}
	
	if user.IsLocked() {
		return nil, errors.New("user account is locked")
	}
	
	if !user.VerifyPassword(password) {
		user.FailedLoginAttempt()
		if err := s.db.Save(user).Error; err != nil {
			return nil, fmt.Errorf("failed to update failed login attempt: %w", err)
		}
		return nil, errors.New("invalid username or password")
	}
	
	// Successful login
	if err := user.Login(); err != nil {
		return nil, fmt.Errorf("login failed: %w", err)
	}
	
	if err := s.db.Save(user).Error; err != nil {
		return nil, fmt.Errorf("failed to update login info: %w", err)
	}
	
	return user, nil
}

// ChangePassword changes a user's password
func (s *UserService) ChangePassword(id uuid.UUID, currentPassword, newPassword string) error {
	user, err := s.GetUserByID(id)
	if err != nil {
		return err
	}
	
	if !user.VerifyPassword(currentPassword) {
		return errors.New("current password is incorrect")
	}
	
	if err := user.SetPassword(newPassword); err != nil {
		return fmt.Errorf("failed to set new password: %w", err)
	}
	
	if err := s.db.Save(user).Error; err != nil {
		return fmt.Errorf("failed to update password: %w", err)
	}
	
	return nil
}

// ResetPassword resets a user's password (admin function)
func (s *UserService) ResetPassword(id uuid.UUID, newPassword string) error {
	user, err := s.GetUserByID(id)
	if err != nil {
		return err
	}
	
	if err := user.SetPassword(newPassword); err != nil {
		return fmt.Errorf("failed to set new password: %w", err)
	}
	
	user.ResetLoginAttempts()
	
	if err := s.db.Save(user).Error; err != nil {
		return fmt.Errorf("failed to update password: %w", err)
	}
	
	return nil
}

// AddPermission adds a permission to a user
func (s *UserService) AddPermission(id uuid.UUID, permission string) error {
	user, err := s.GetUserByID(id)
	if err != nil {
		return err
	}
	
	user.AddPermission(permission)
	
	if err := s.db.Save(user).Error; err != nil {
		return fmt.Errorf("failed to add permission: %w", err)
	}
	
	return nil
}

// RemovePermission removes a permission from a user
func (s *UserService) RemovePermission(id uuid.UUID, permission string) error {
	user, err := s.GetUserByID(id)
	if err != nil {
		return err
	}
	
	user.RemovePermission(permission)
	
	if err := s.db.Save(user).Error; err != nil {
		return fmt.Errorf("failed to remove permission: %w", err)
	}
	
	return nil
}

// ExportUsers exports users to JSON
func (s *UserService) ExportUsers() ([]byte, error) {
	users, _, err := s.GetAllUsers(1, 1000) // Get all users (limit to 1000 for safety)
	if err != nil {
		return nil, fmt.Errorf("failed to get users for export: %w", err)
	}
	
	var responses []*models.UserResponse
	for _, user := range users {
		responses = append(responses, user.ToResponse())
	}
	
	data, err := json.MarshalIndent(responses, "", "  ")
	if err != nil {
		return nil, fmt.Errorf("failed to marshal users: %w", err)
	}
	
	return data, nil
}

// GetUserActivity returns user activity information
func (s *UserService) GetUserActivity(id uuid.UUID) (*utils.UserActivity, error) {
	user, err := s.GetUserByID(id)
	if err != nil {
		return nil, err
	}
	
	activity := &utils.UserActivity{
		UserID:        user.ID,
		Username:      user.Username,
		LastLogin:     user.LastLogin,
		LoginAttempts: user.LoginAttempts,
		IsActive:      user.IsActive(),
		IsLocked:      user.IsLocked(),
		CreatedAt:     user.CreatedAt,
		UpdatedAt:     user.UpdatedAt,
	}
	
	return activity, nil
}

File: test/sample-projects/typescript/user-management/src/services/UserService.ts
import { User } from '../models/User';
import {
  IUser,
  IUserResponse,
  ICreateUser,
  IUpdateUser,
  IUserStats,
  IAuthResult,
  IUserActivity,
  IPaginatedUsers,
  ISearchUsersResponse,
  IUserFilter,
  UserRole,
} from '../types/User';
import { AppError } from '../utils/errors';
import logger from '../utils/logger';

/**
 * UserService class handles all user-related business logic
 */
export class UserService {
  /**
   * Create a new user
   * @param userData - User data object
   * @returns Created user response
   */
  async createUser(userData: ICreateUser): Promise<IUserResponse> {
    try {
      // Check if username already exists
      const existingUsername = await User.findByUsername(userData.username);
      if (existingUsername) {
        throw new AppError('Username already exists', 400);
      }

      // Check if email already exists (if provided)
      if (userData.email) {
        const existingEmail = await User.findByEmail(userData.email);
        if (existingEmail) {
          throw new AppError('Email already exists', 400);
        }
      }

      // Create new user
      const user = new User(userData);
      
      // Validate user data
      const validationErrors = user.validateUser();
      if (validationErrors.length > 0) {
        throw new AppError(validationErrors.join(', '), 400);
      }

      await user.save();
      
      logger.info(`User created successfully: ${user.username}`);
      return user.response;
    } catch (error) {
      logger.error('Error creating user:', error as Error);
      throw error;
    }
  }

  /**
   * Get user by ID
   * @param id - User ID
   * @returns User response
   */
  async getUserById(id: string): Promise<IUserResponse> {
    try {
      const user = await User.findOne({ id });
      if (!user) {
        throw new AppError('User not found', 404);
      }
      return user.response;
    } catch (error) {
      logger.error('Error getting user by ID:', error as Error);
      throw error;
    }
  }

  /**
   * Get user by username
   * @param username - Username
   * @returns User response
   */
  async getUserByUsername(username: string): Promise<IUserResponse> {
    try {
      const user = await User.findByUsername(username);
      if (!user) {
        throw new AppError('User not found', 404);
      }
      return user.response;
    } catch (error) {
      logger.error('Error getting user by username:', error as Error);
      throw error;
    }
  }

  /**
   * Get user by email
   * @param email - Email address
   * @returns User response
   */
  async getUserByEmail(email: string): Promise<IUserResponse> {
    try {
      const user = await User.findByEmail(email);
      if (!user) {
        throw new AppError('User not found', 404);
      }
      return user.response;
    } catch (error) {
      logger.error('Error getting user by email:', error as Error);
      throw error;
    }
  }

  /**
   * Update user
   * @param id - User ID
   * @param updateData - Update data
   * @returns Updated user response
   */
  async updateUser(id: string, updateData: IUpdateUser): Promise<IUserResponse> {
    try {
      const user = await User.findOne({ id });
      if (!user) {
        throw new AppError('User not found', 404);
      }

      // Apply updates (excluding password and id)
      Object.keys(updateData).forEach(key => {
        if (key !== 'password' && key !== 'id') {
          (user as any)[key] = (updateData as any)[key];
        }
      });

      // Validate updated data
      const validationErrors = user.validateUser();
      if (validationErrors.length > 0) {
        throw new AppError(validationErrors.join(', '), 400);
      }

      await user.save();
      
      logger.info(`User updated successfully: ${user.username}`);
      return user.response;
    } catch (error) {
      logger.error('Error updating user:', error as Error);
      throw error;
    }
  }

  /**
   * Delete user (soft delete)
   * @param id - User ID
   * @returns Success status
   */
  async deleteUser(id: string): Promise<boolean> {
    try {
      const user = await User.findOne({ id });
      if (!user) {
        throw new AppError('User not found', 404);
      }

      user.delete();
      await user.save();
      
      logger.info(`User deleted successfully: ${user.username}`);
      return true;
    } catch (error) {
      logger.error('Error deleting user:', error as Error);
      throw error;
    }
  }

  /**
   * Hard delete user (permanent deletion)
   * @param id - User ID
   * @returns Success status
   */
  async hardDeleteUser(id: string): Promise<boolean> {
    try {
      const result = await User.deleteOne({ id });
      if (result.deletedCount === 0) {
        throw new AppError('User not found', 404);
      }
      
      logger.info(`User permanently deleted: ${id}`);
      return true;
    } catch (error) {
      logger.error('Error hard deleting user:', error as Error);
      throw error;
    }
  }

  /**
   * Get all users with pagination
   * @param page - Page number
   * @param limit - Items per page
   * @param filter - Filter criteria
   * @returns Paginated users response
   */
  async getAllUsers(
    page: number = 1,
    limit: number = 20,
    filter: IUserFilter = {}
  ): Promise<IPaginatedUsers> {
    try {
      const skip = (page - 1) * limit;
      
      const users = await User.find(filter)
        .sort({ createdAt: -1 })
        .skip(skip)
        .limit(limit);

      const total = await User.countDocuments(filter);
      const totalPages = Math.ceil(total / limit);

      return {
        users: users.map(user => user.response),
        pagination: {
          page,
          limit,
          total,
          totalPages,
          hasNext: page < totalPages,
          hasPrev: page > 1,
        },
      };
    } catch (error) {
      logger.error('Error getting all users:', error as Error);
      throw error;
    }
  }

  /**
   * Get active users
   * @returns Active users
   */
  async getActiveUsers(): Promise<IUserResponse[]> {
    try {
      const users = await User.findActive();
      return users.map(user => user.response);
    } catch (error) {
      logger.error('Error getting active users:', error as Error);
      throw error;
    }
  }

  /**
   * Get users by role
   * @param role - User role
   * @returns Users with specified role
   */
  async getUsersByRole(role: UserRole): Promise<IUserResponse[]> {
    try {
      const users = await User.findByRole(role);
      return users.map(user => user.response);
    } catch (error) {
      logger.error('Error getting users by role:', error as Error);
      throw error;
    }
  }

  /**
   * Search users
   * @param query - Search query
   * @param page - Page number
   * @param limit - Items per page
   * @returns Search results
   */
  async searchUsers(
    query: string,
    page: number = 1,
    limit: number = 20
  ): Promise<ISearchUsersResponse> {
    try {
      const skip = (page - 1) * limit;
      
      const users = await User.searchUsers(query, {
        skip,
        limit,
        sort: { createdAt: -1 },
      });

      // Count total matching users
      const totalUsers = await User.searchUsers(query);
      const total = totalUsers.length;
      const totalPages = Math.ceil(total / limit);

      return {
        users: users.map(user => user.response),
        query,
        pagination: {
          page,
          limit,
          total,
          totalPages,
          hasNext: page < totalPages,
          hasPrev: page > 1,
        },
      };
    } catch (error) {
      logger.error('Error searching users:', error as Error);
      throw error;
    }
  }

  /**
   * Get user statistics
   * @returns User statistics
   */
  async getUserStats(): Promise<IUserStats> {
    try {
      const stats = await User.getUserStats();
      return stats;
    } catch (error) {
      logger.error('Error getting user statistics:', error as Error);
      throw error;
    }
  }

  /**
   * Authenticate user
   * @param username - Username
   * @param password - Password
   * @returns Authentication result
   */
  async authenticateUser(username: string, password: string): Promise<IAuthResult> {
    try {
      const user = await User.findByUsername(username).select('+password');
      
      if (!user || !(await user.checkPassword(password))) {
        // Record failed login attempt if user exists
        if (user) {
          user.recordFailedLogin();
          await user.save();
        }
        throw new AppError('Invalid username or password', 401);
      }

      if (!user.isActive) {
        throw new AppError('User account is not active', 401);
      }

      if (user.isLocked) {
        throw new AppError('User account is locked', 401);
      }

      // Record successful login
      user.recordLogin();
      await user.save();

      // Generate token
      const token = user.generateToken();

      logger.info(`User authenticated successfully: ${user.username}`);
      
      return {
        user: user.response,
        token,
      };
    } catch (error) {
      logger.error('Error authenticating user:', error as Error);
      throw error;
    }
  }

  /**
   * Change user password
   * @param id - User ID
   * @param currentPassword - Current password
   * @param newPassword - New password
   * @returns Success status
   */
  async changePassword(
    id: string,
    currentPassword: string,
    newPassword: string
  ): Promise<boolean> {
    try {
      const user = await User.findOne({ id }).select('+password');
      if (!user) {
        throw new AppError('User not found', 404);
      }

      if (!(await user.checkPassword(currentPassword))) {
        throw new AppError('Current password is incorrect', 400);
      }

      user.password = newPassword;
      await user.save();

      logger.info(`Password changed successfully for user: ${user.username}`);
      return true;
    } catch (error) {
      logger.error('Error changing password:', error as Error);
      throw error;
    }
  }

  /**
   * Reset user password (admin function)
   * @param id - User ID
   * @param newPassword - New password
   * @returns Success status
   */
  async resetPassword(id: string, newPassword: string): Promise<boolean> {
    try {
      const user = await User.findOne({ id });
      if (!user) {
        throw new AppError('User not found', 404);
      }

      user.password = newPassword;
      user.resetLoginAttempts();
      await user.save();

      logger.info(`Password reset successfully for user: ${user.username}`);
      return true;
    } catch (error) {
      logger.error('Error resetting password:', error as Error);
      throw error;
    }
  }

  /**
   * Add permission to user
   * @param id - User ID
   * @param permission - Permission to add
   * @returns Success status
   */
  async addPermission(id: string, permission: string): Promise<boolean> {
    try {
      const user = await User.findOne({ id });
      if (!user) {
        throw new AppError('User not found', 404);
      }

      user.addPermission(permission);
      await user.save();

      logger.info(`Permission added to user ${user.username}: ${permission}`);
      return true;
    } catch (error) {
      logger.error('Error adding permission:', error as Error);
      throw error;
    }
  }

  /**
   * Remove permission from user
   * @param id - User ID
   * @param permission - Permission to remove
   * @returns Success status
   */
  async removePermission(id: string, permission: string): Promise<boolean> {
    try {
      const user = await User.findOne({ id });
      if (!user) {
        throw new AppError('User not found', 404);
      }

      user.removePermission(permission);
      await user.save();

      logger.info(`Permission removed from user ${user.username}: ${permission}`);
      return true;
    } catch (error) {
      logger.error('Error removing permission:', error as Error);
      throw error;
    }
  }

  /**
   * Export users data
   * @returns Users data for export
   */
  async exportUsers(): Promise<IUserResponse[]> {
    try {
      const users = await User.find().sort({ createdAt: -1 });
      return users.map(user => user.response);
    } catch (error) {
      logger.error('Error exporting users:', error as Error);
      throw error;
    }
  }

  /**
   * Get user activity
   * @param id - User ID
   * @returns User activity data
   */
  async getUserActivity(id: string): Promise<IUserActivity> {
    try {
      const user = await User.findOne({ id });
      if (!user) {
        throw new AppError('User not found', 404);
      }

      return {
        id: user.id,
        username: user.username,
        lastLogin: user.lastLogin,
        loginAttempts: user.loginAttempts,
        isActive: user.isActive,
        isLocked: user.isLocked,
        createdAt: user.createdAt,
        updatedAt: user.updatedAt,
      };
    } catch (error) {
      logger.error('Error getting user activity:', error as Error);
      throw error;
    }
  }
}

export default UserService;